BlockStorage: use serializer for Signature in updateBlockSig

omerfirmak opened the pull request:
Blocks are deserialized as a whole. Instead of toBlob(), updated signatures
should use serializer too.

Fixes #1848

Geod24 commented:
But we still have to overwrite *some* data to know where to find the new sigs, no ?

omerfirmak commented:
> But we still have to overwrite _some_ data to know where to find the new sigs, no ?

Yes, but it is a `size_t` at a known position.

Geod24 commented:
So a fixed-size data structure at a known position... Like the signature was supposed to be :O

omerfirmak committed:
BlockStorage: Fix readSizeT passing wrong length to read

omerfirmak commented:
> So a fixed-size data structure at a known position... Like the signature was supposed to be :O

`validators` BitField is probably the problematic one.

Also , fixed the CI failure.

hewison-chris commented:
> > So a fixed-size data structure at a known position... Like the signature was supposed to be :O
> 
> `validators` BitField is probably the problematic one.
> 
> Also , fixed the CI failure.

The `validators` BitField is also fixed size for the block height as it is determined by the number of active validators which can not change once the previous block is externalized.

Geod24 commented:
Also we re-order the block header to make sure the signature had a fixed offset: https://github.com/bosagora/agora/commit/4ecafac7de24a460e7609a3f5fa388229484457a

Geod24 commented:
Can you add a test that triggers the bug in https://github.com/bosagora/agora/tree/v0.x.x/tests/unit ?

omerfirmak commented:
> > > So a fixed-size data structure at a known position... Like the signature was supposed to be :O
> > 
> > 
> > `validators` BitField is probably the problematic one.
> > Also , fixed the CI failure.
> 
> The `validators` BitField is also fixed size for the block height as it is determined by the number of active validators which can not change once the previous block is externalized.

The question is if the serialized `validators` changes in size or not. I think our `ubyte[]` serialization was not that smart, there was an issue for it IIRC.

Geod24 commented on source/agora/node/BlockStorage.d:
```diff
@@ -568,7 +574,7 @@ public class BlockStorage : IBlockStorage
     private size_t readSizeT (size_t pos) @trusted
     {
-        ubyte[] data = this.read(pos, pos+size_t.sizeof);
+        ubyte[] data = this.read(pos, size_t.sizeof);
```
This seems to be the bug, no ?

omerfirmak replied:
I don't think it is "the" bug we are looking for.

ferencdg commented:
> > > > So a fixed-size data structure at a known position... Like the signature was supposed to be :O
> > > 
> > > 
> > > `validators` BitField is probably the problematic one.
> > > Also , fixed the CI failure.
> > 
> > 
> > The `validators` BitField is also fixed size for the block height as it is determined by the number of active validators which can not change once the previous block is externalized.
> 
> The question is if the serialized `validators` changes in size or not. I think our `ubyte[]` serialization was not that smart, there was an issue for it IIRC.

I quickly checked the ubyte array serialization, but it looked okay.

I wonder if the issue is that we are relying on those flaky in-memory maps(index_to_key) to get the number of validators in a lot of places in our code, and later we use that to initialize the Bitfield. If those maps are not properly initialized(and have 0 validators for a particular height), then our Bitfield will have 0 size, and it will be 1 byte shorter in the serialized form.

ferencdg commented:
valueExternalized calls updateMultiSignature calls getCountOfValidators uses key_to_index map for the new block height....
isn't that map updated for height only after we added the block to the blockchain? That would explain why the serialized form of the first block is 1 byte shorter

ferencdg commented:
updateMultiSignature calls updateMultiSignature calls getCountOfValidators uses index_to_key for the new height. I think index_to_key is only updated for the new height after the block is added to the blockchain and that explains why the serialized form is 1 byte shorter the first time we add the block to the storage.

ferencdg commented:
valueExternalized calls updateMultiSignature calls getCountOfValidators uses index_to_key for the the new height. I think that map is only updated for the new height after the block is added to the chain, and that explains why the serialized form is 1 byte shorter the first time we add the block to the block storage.

omerfirmak committed:
BlockStorage: use serializer for Signature in updateBlockSig

Blocks are deserialized as a whole. Instead of toBlob(), updated signatures
should use serializer too.

omerfirmak commented:
Ok, the real issue was MUCH simpler. See the updated commits.

Geod24 commented:
Noice! I have no clue why I wrote this code. Is it realistic to ask for a test ?

omerfirmak commented:
> Noice! I have no clue why I wrote this code. Is it realistic to ask for a test ?

The existing `BlockStorageMultiSig` would have caught this if the Signatures were not repeated deadbeef. Lets just make them random for each block?

omerfirmak committed:
Test: Use random sigs in BlockStorageMultiSig

omerfirmak commented:
> > Noice! I have no clue why I wrote this code. Is it realistic to ask for a test ?
> 
> The existing `BlockStorageMultiSig` would have caught this if the Signatures were not repeated deadbeef. Lets just make them random for each block?

Done
